src/Core/Content/MailTemplate/Service/MailTemplateService.php

```diff
@@ -0,0 +1,241 @@
+
+    /**
+     * @internal
+     */
+    public function __construct(
+        Environment $twig,
+        private readonly TwigVariableParserFactory $twigVariableParserFactory,
+        private readonly DefinitionInstanceRegistry $definitionRegistry,
+        private readonly DataValidator $dataValidator,
+    ) {
+        $this->twigVariableParser = $this->twigVariableParserFactory->getParser($twig);
+    }
+
+    /**
+     * @return MailTemplateValidationResponse[]
+     */
+    public function validateTemplate(string $mailTemplate, EventDataCollection $availableVariables): array
+    {
+        try {
+            $usedVariables = $this->twigVariableParser->parse($mailTemplate);
+        } catch (SyntaxError $exception) {
+            return [new MailTemplateValidationError(
+                $this->dataValidator,
+                MailTemplateValidationError::TYPE_SYNTAX,
+                ['message' => $exception->getRawMessage()],
+                $exception->getTemplateLine(),
+            )];
+        }
+
+        return $this->validateVariables($usedVariables, $availableVariables);
+    }
+
+    /**
+     * @param array<string, string> $usedVariables
+     *
+     * @return MailTemplateValidationResponse[]
+     */
+    public function validateVariables(array $usedVariables, EventDataCollection $availableVariables): array
+    {
+        $errors = [];
+
+        foreach ($usedVariables as $var) {
+            if ($this->isKnownVariable($var, $availableVariables, $errors)) {
+                continue;
+            }
+
+            $this->validateComplexVariable($var, $availableVariables, $errors);
+        }
+
+        return $errors;
+    }
+
+    /**
+     * @param MailTemplateValidationResponse[] $errors
+     */
+    private function isKnownVariable(string $var, EventDataCollection $availableVariables, array &$errors): bool
+    {
+        if ($field = $availableVariables->get($var)) {
+            if (!($field instanceof ScalarValueType)) {
+                $errors[] = new MailTemplateValidationWarning(
+                    $this->dataValidator,
+                    MailTemplateValidationWarning::TYPE_COMPLEX_ELEMENT,
+                    ['variable' => $var]
+                );
+            }
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * @param MailTemplateValidationResponse[] $errors
+     */
+    private function validateComplexVariable(string $var, EventDataCollection $availableVariables, array &$errors): void
+    {
+        if (!str_contains($var, '.')) {
+            $errors[] = new MailTemplateValidationError(
+                $this->dataValidator,
+                MailTemplateValidationError::TYPE_UNKNOWN_VARIABLE,
+                ['variable' => $var]
+            );
+            return;
+        }
+
+        $varParts = explode('.', $var);
+
+        $nestedAvVars = $availableVariables->get($varParts[0]);
+
+        for ($i = 0; $i < \count($varParts); ++$i) {
```

Comment:
you should add a variable for the count before the loop. otherwise it will be calculated for each loop